app/routes/login.tsx

```diff
@@ -0,0 +1,72 @@
+import { Center, VStack, Heading, Text } from "@chakra-ui/react";
+import { withZod } from "@remix-validated-form/with-zod";
+import { ActionFunction, json, useActionData } from "remix";
+import { ValidatedForm, validationError } from "remix-validated-form";
+import { zfd } from "zod-form-data";
+import { createUserSession, verifyPassword } from "~app/auth";
+import FormInput from "~app/form/FormInput";
+import SubmitButton from "~app/form/SubmitButton";
+import { PrismaClient as prisma } from "~app/prisma";
+
+const validator = withZod(
+	zfd.formData({
+		password: zfd.text(),
+		username: zfd.text(),
+	}),
+);
+
+export const action: ActionFunction = async ({ request }) => {
+	const fieldValues = await validator.validate(await request.formData());
+	if (fieldValues.error) return validationError(fieldValues.error);
+	const { username, password } = fieldValues.data;
+
+	const user = await prisma.user.findUnique({ where: { username } });
+
+	if (
+		!verifyPassword(
+			password,
+			user?.password ??
+				// cspell:disable-next-line
+				"$2a$10$ctXCCuxt.yedpcwAS1kaw.xcC/OniG.FGOayPPEU9KKppw1wJGHGW",
+		)
+	) {
+		return json({ formError: "Password or username is incorrect" }, 400);
+	}
```

Comment:
This is missing `needsRehash` and other checks from the previous Next.js application ([source](https://github.com/3x071c/lsg/blob/7c377cdef5baddd9dcf5f49985325081f31159e3/graphql/schema/User.ts#L100-L132))

Reply:
What exactly is `needsRehash` for?

Reply:
@Flosi23 `needsRehash` checks if the hash is outdated (configuration changes etc.)